### PR TITLE
Check for literals as println! args

### PR DESCRIPTION
**Pull request by mrecachinas**
Addresses #2480

### CODE REVIEW

File: clippy_lints/src/print.rs [hunk 1]
```diff
@@ -78,12 +78,28 @@ declare_lint! {
...
+///
+/// **Why is this bad?** Using literals as `println!` args is inefficient
+/// (c.f., https://github.com/matthiaskrgr/rust-str-bench) and unnecessary
+/// (i.e., just put the literal in the format string)
+///
+/// **Example:**
+/// ```rust
+/// println!("{}", "foo");
+/// ```
+declare_lint! {
```
Comment (oli-obk) [inline clippy_lints/src/print.rs]: We're deprecating the use of this macro. Please use `declare_clippy_lint` together with the nursery group

**Comment by mrecachinas** [comment]
Good catch on the `println!("foo: {}", env!("BAR"));` -- it does throw the warning. I'll take a look at that and modify to use the `declare_clippy_lint`. 

Currently running into some weirdness when running `update_all_references.sh`. It's modifying tests/ui/print.rs, tests/ui/print_with_newline.rs, and tests/ui/format.rs even with `#![allow(print_literal)]`.

**Comment by oli-obk** [comment]
Sometimes you need to rerun all tests, run the script, run tests again, run the script before the update script catches on. This also happens inside rustc... no clue what's exactly happening.

**Comment by oli-obk** [comment]
You can probably fix the env! issue with very targetted in_macro calls. Not sure how that'll work, since println is also a macro. But technically the args should not be inside a macro expansion

**Comment by mrecachinas** [comment]
Do you think this should be pulled out of print.rs or run as an `EarlyLintPass`? I'm not seeing any indicator of a macro call (in this case to `env!`).

File: clippy_lints/src/print.rs (hunk 1 above)
Comment (mrecachinas) [reply clippy_lints/src/print.rs]: This is fixed in 511aa65.

**Comment by oli-obk** [comment]
Ah, thanks for investigating. Can you just list this issue under the downsides of the lint? It's fine to have this weird corner case imo

**Comment by mrecachinas** [comment]
Sounds good. I added the test-case and added it to "Known Problems."

Let me know if there's anything else you'd like to see in this.

**Comment by mrecachinas** [comment]
To make the tests reproducible (on all platforms) (and without explicitly setting an environment var), I removed the test with the `env!`. I suppose a `format!` would accomplish the same thing. Let me know if you want it added.

**Comment by oli-obk** [comment]
Nope, this is fine as it is. Thanks
